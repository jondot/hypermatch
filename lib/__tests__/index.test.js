'use strict';

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var run = require('../');

var t = ['or', ['or', ['includes', 'name', ['a']], ['equals', 'browser', 'firefox']], ['includes', 'uid', ['user1']]];

var t1 = ['and', ['includes', 'name', ['a']], ['equals', 'browser', 'firefox']];

var bad = ['and', ['incluuudes', 'name', ['a']], ['equals', 'browser', 'firefox']];

var t3 = ['not', ['equals', 'browser', 'firefox']];

var r1 = ['and', ['equals', 'uid', '1'], ['regex', 'name', /moo.*/]];

var Tracer = function () {
  function Tracer() {
    (0, _classCallCheck3.default)(this, Tracer);

    this.actions = [];
    this.log = this.log.bind(this);
  }

  (0, _createClass3.default)(Tracer, [{
    key: 'log',
    value: function log() {
      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      this.actions.push(args);
    }
  }]);
  return Tracer;
}();

var testRun = function testRun(res, tree, props) {
  test((0, _stringify2.default)(tree) + ' ' + (res ? '✓' : '✕') + ' ' + (0, _stringify2.default)(props), function () {
    var tracer = new Tracer();
    expect(run(tree, props, { trace: true, log: tracer.log })).toEqual(res);
    expect(tracer.actions).toMatchSnapshot();
  });
};

test('handles bad syntax', function () {
  return expect(function () {
    run(bad, { name: 'a' });
  }).toThrow();
});

testRun(false, t1, { uid: 'doesnt', name: 'match' });
testRun(true, t1, { name: 'a', browser: 'firefox' });
testRun(false, t1, { name: 'a', browser: 'nope!' });

testRun(true, t, { uid: 'user1', name: 'b', browser: 'android' });

testRun(true, t3, { browser: 'android' });
testRun(false, t3, { browser: 'firefox' });

testRun(true, ['equals', 'name', 'foobar'], { name: 'foobar' });
testRun(false, ['equals', 'name', 'foobar'], { name: 'foo' });

testRun(true, ['includes', 'name', ['m1', 'm2']], { name: 'm2' });
testRun(false, ['includes', 'name', ['m1', 'm2']], { name: 'm3' });

testRun(false, t, { name: 'bad', browser: 'foobar', uid: 'poop' });

// regex
testRun(true, ['regex', 'name', 'moo.*'], { name: 'moomoo moo!', uid: '1' });
testRun(true, r1, { name: 'moomoo moo!', uid: '1' });
testRun(false, r1, { name: 'poop', uid: '1' });

// subset
testRun(true, ['subset', 'orderIds', ['O-1', 'O-2', 'O-3', 'O-4']], {
  orderIds: ['O-1', 'O-2', 'O-3']
});
testRun(false, ['subset', 'orderIds', ['O-1', 'O-2', 'O-3']], {
  orderIds: ['O-2', 'O-3', 'O-4']
});
testRun(true, ['subset', 'orderIds', ['O-1', 'O-2', 'O-3']], { orderIds: [] });
testRun(true, ['subset', 'orderIds', []], { orderIds: [] });

// intersects
testRun(true, ['intersects', 'professions', ['barista', 'baker', 'bartender', 'waiter']], {
  professions: ['chef', 'baker', 'bartender']
});
testRun(false, ['intersects', 'professions', ['baker', 'barista']], { 'professions': ['waiter'] });
testRun(true, ['intersects', 'professions', ['barista']], { professions: ['barista'] });

// range
testRun(true, ['range', 'age', [18, 40]], { age: 23 });
testRun(false, ['range', 'age', [18, 40]], { age: 66 });

// exists
testRun(true, ['exists', 'age'], { age: 23 });
testRun(false, ['exists', 'age'], { foobar: 23 });

// lt, gt
testRun(false, ['lt', 'age', 12], { age: 23 });
testRun(true, ['gt', 'age', 12], { age: 23 });
testRun(true, ['lt', 'age', 40], { age: 23 });
testRun(false, ['gt', 'age', 40], { age: 23 });