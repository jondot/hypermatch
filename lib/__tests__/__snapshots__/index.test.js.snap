// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`["and",["equals","uid","1"],["regex","name",{}]] ✓ {"name":"moomoo moo!","uid":"1"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "equals",
        "uid",
        "1",
      ],
      Array [
        "regex",
        "name",
        /moo\\.\\*/,
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "1",
      "1",
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "regex",
    Array [
      /moo\\.\\*/,
      "moomoo moo!",
    ],
  ],
]
`;

exports[`["and",["equals","uid","1"],["regex","name",{}]] ✕ {"name":"poop","uid":"1"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "equals",
        "uid",
        "1",
      ],
      Array [
        "regex",
        "name",
        /moo\\.\\*/,
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "1",
      "1",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "regex",
    Array [
      /moo\\.\\*/,
      "poop",
    ],
  ],
]
`;

exports[`["and",["includes","name",["a"]],["equals","browser","firefox"]] ✓ {"name":"a","browser":"firefox"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "a",
      ],
      "a",
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "firefox",
      "firefox",
    ],
  ],
]
`;

exports[`["and",["includes","name",["a"]],["equals","browser","firefox"]] ✕ {"name":"a","browser":"nope!"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "a",
      ],
      "a",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "nope!",
    ],
  ],
]
`;

exports[`["and",["includes","name",["a"]],["equals","browser","firefox"]] ✕ {"uid":"doesnt","name":"match"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "and",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "a",
      ],
      "match",
    ],
  ],
]
`;

exports[`["equals","name","foobar"] ✓ {"name":"foobar"} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "foobar",
      "foobar",
    ],
  ],
]
`;

exports[`["equals","name","foobar"] ✕ {"name":"foo"} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "foobar",
      "foo",
    ],
  ],
]
`;

exports[`["exists","age"] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "exists",
    Array [
      23,
    ],
  ],
]
`;

exports[`["exists","age"] ✕ {"foobar":23} 1`] = `Array []`;

exports[`["gt","age",12] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "gt",
    Array [
      12,
      23,
    ],
  ],
]
`;

exports[`["gt","age",40] ✕ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "gt",
    Array [
      40,
      23,
    ],
  ],
]
`;

exports[`["includes","name",["m1","m2"]] ✓ {"name":"m2"} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "m1",
        "m2",
      ],
      "m2",
    ],
  ],
]
`;

exports[`["includes","name",["m1","m2"]] ✕ {"name":"m3"} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "m1",
        "m2",
      ],
      "m3",
    ],
  ],
]
`;

exports[`["intersects","professions",["baker","barista"]] ✕ {"professions":["waiter"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "intersects",
    Array [
      Array [
        "baker",
        "barista",
      ],
      Array [
        "waiter",
      ],
    ],
  ],
]
`;

exports[`["intersects","professions",["barista","baker","bartender","waiter"]] ✓ {"professions":["chef","baker","bartender"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "intersects",
    Array [
      Array [
        "barista",
        "baker",
        "bartender",
        "waiter",
      ],
      Array [
        "chef",
        "baker",
        "bartender",
      ],
    ],
  ],
]
`;

exports[`["intersects","professions",["barista"]] ✓ {"professions":["barista"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "intersects",
    Array [
      Array [
        "barista",
      ],
      Array [
        "barista",
      ],
    ],
  ],
]
`;

exports[`["lt","age",12] ✕ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "lt",
    Array [
      12,
      23,
    ],
  ],
]
`;

exports[`["lt","age",40] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "lt",
    Array [
      40,
      23,
    ],
  ],
]
`;

exports[`["not",["equals","browser","firefox"]] ✓ {"browser":"android"} 1`] = `
Array [
  Array [
    "op[1]  ",
    "not",
    Array [
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "android",
    ],
  ],
]
`;

exports[`["not",["equals","browser","firefox"]] ✕ {"browser":"firefox"} 1`] = `
Array [
  Array [
    "op[1]  ",
    "not",
    Array [
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "equals",
    Array [
      "firefox",
      "firefox",
    ],
  ],
]
`;

exports[`["or",["or",["includes","name",["a"]],["equals","browser","firefox"]],["includes","uid",["user1"]]] ✓ {"uid":"user1","name":"b","browser":"android"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "or",
        Array [
          "includes",
          "name",
          Array [
            "a",
          ],
        ],
        Array [
          "equals",
          "browser",
          "firefox",
        ],
      ],
      Array [
        "includes",
        "uid",
        Array [
          "user1",
        ],
      ],
    ],
  ],
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "a",
      ],
      "b",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "android",
    ],
  ],
  Array [
    "op[.]  ",
    true,
    "includes",
    Array [
      Array [
        "user1",
      ],
      "user1",
    ],
  ],
]
`;

exports[`["or",["or",["includes","name",["a"]],["equals","browser","firefox"]],["includes","uid",["user1"]]] ✕ {"name":"bad","browser":"foobar","uid":"poop"} 1`] = `
Array [
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "or",
        Array [
          "includes",
          "name",
          Array [
            "a",
          ],
        ],
        Array [
          "equals",
          "browser",
          "firefox",
        ],
      ],
      Array [
        "includes",
        "uid",
        Array [
          "user1",
        ],
      ],
    ],
  ],
  Array [
    "op[>]  ",
    "or",
    Array [
      Array [
        "includes",
        "name",
        Array [
          "a",
        ],
      ],
      Array [
        "equals",
        "browser",
        "firefox",
      ],
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "a",
      ],
      "bad",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "equals",
    Array [
      "firefox",
      "foobar",
    ],
  ],
  Array [
    "op[.]  ",
    false,
    "includes",
    Array [
      Array [
        "user1",
      ],
      "poop",
    ],
  ],
]
`;

exports[`["range","age",[18,40]] ✓ {"age":23} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "range",
    Array [
      Array [
        18,
        40,
      ],
      23,
    ],
  ],
]
`;

exports[`["range","age",[18,40]] ✕ {"age":66} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "range",
    Array [
      Array [
        18,
        40,
      ],
      66,
    ],
  ],
]
`;

exports[`["regex","name","moo.*"] ✓ {"name":"moomoo moo!","uid":"1"} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "regex",
    Array [
      "moo.*",
      "moomoo moo!",
    ],
  ],
]
`;

exports[`["subset","orderIds",["O-1","O-2","O-3","O-4"]] ✓ {"orderIds":["O-1","O-2","O-3"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "subset",
    Array [
      Array [
        "O-1",
        "O-2",
        "O-3",
        "O-4",
      ],
      Array [
        "O-1",
        "O-2",
        "O-3",
      ],
    ],
  ],
]
`;

exports[`["subset","orderIds",["O-1","O-2","O-3"]] ✓ {"orderIds":[]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "subset",
    Array [
      Array [
        "O-1",
        "O-2",
        "O-3",
      ],
      Array [],
    ],
  ],
]
`;

exports[`["subset","orderIds",["O-1","O-2","O-3"]] ✕ {"orderIds":["O-2","O-3","O-4"]} 1`] = `
Array [
  Array [
    "op[.]  ",
    false,
    "subset",
    Array [
      Array [
        "O-1",
        "O-2",
        "O-3",
      ],
      Array [
        "O-2",
        "O-3",
        "O-4",
      ],
    ],
  ],
]
`;

exports[`["subset","orderIds",[]] ✓ {"orderIds":[]} 1`] = `
Array [
  Array [
    "op[.]  ",
    true,
    "subset",
    Array [
      Array [],
      Array [],
    ],
  ],
]
`;
